enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,164 @@
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will be populated in every cluster to provide available CLIs and plugins for a cluster.
+
+### Non-Goals
+
+* CLI manager will not build or serve the binaries.  It will know where to find them.
+* This proposal is not concerned with _which_ binaries will be managed.  This proposal is meant to determine the mechanism only.  Consumers and publishers are clients of the mechanism. 
+* CLI manager will not create or update the ClusterCLI Custom Resources, those will be managed by individual CLIs and their operators. 
+
+### Requirements:
+
+1. CLI author requirements:
+    - provide CLI image containing binaries
+    - Custom Resource that describes the binaries (location within the image, metadata, description)
+2. An operator that will be responsible for:
+    - registering the new CLI CRD 
+    - reading metadata from CR from authors and extracting the binaries from the CLI images to local disk
+    - serving extracted binaries
+    - serving krew-compatible index
+3. oc changes:
+    - embed krew
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user, I want a CLI in-cluster manager for various CLIs available for OpenShift and related services so that I can discover, install and list them
+The user will invoke the following commands:
+
+* `oc krew list` to view installed plugins
+* `oc krew search` will communicate with the krew index served by the cli-in-cluster-manager operator and will retrieve information about CLIs
+* `oc krew install foo` will install `foo` to a user's system.
+
+##### example:
+```
+$ oc krew list
+NAME                 DESC                        LATEST       INSTALLED
+-----                -----                       -----        -----
+openshift-installer  OpenShift Installer CLI     4.4          4.3
+kubectl              Kubernetes CLI              1.15         1.13
+oc                   OpenShift CLI               4.4          4.3
+odo                  OpenShift Developer CLI     1.0          Not Installed
+kn                   Knative CLI                 1.0          Not Installed
+tkn                  Tekton CLI                  1.0          Not Installed
+
+```
+
+## Design Detals
+
+Each CLI will provide a separate image.
+Each CLI is responsible for creating a CR to hold metadata that `oc krew` will depend on.  The CR will serve to deliver the metadata and description
+of its deliverable binary.  An in-cluster CLI manager controller will access the CRs to populate and maintain an in-cluster Krew Index.  Users will 
+install OpenShift tools that are known compatible with each cluster version through `oc krew`.    
+
+A CLI must provide a ClusterCLI CR.  The result of this proposal will be:
+* ClusterCLI Custom Resource Definition
+* Krew Index and a controller to create the index from CLI CRs.
```

Comment:
Why does the index have to be created on the server?  If you create a binary distribution mechanism in cluster you have to get security audit and review on third party code there - it might be better to create the index locally in oc from CRs on cluster.

Reply:
The index will be kept locally, we'd be moving the extraction mechanism which currently is part of `oc adm release extract --tools` from oc to the server and providing the binaries extracted this way through a web server. Not sure if this qualifies as a distribution mechanism, since the distribution is just like what we do today - images. We're only moving the extraction from client (currently only oc knows how to handle that) to the server, thus allowing other consumers to download these binaries.